### PR TITLE
Add support for checking fields values

### PR DESCRIPTION
**Pull request by mirzal**
This is follow-up to PR #141 . Main changes compared to that one:

- drop `validate-values` command - do values check as part of `validate`
- changed global SETTINGS structure - now we have `nocolor`, `json` and `tokens`, where most of fun is. `tokens` is dictionary keyed by token name, values are `TokenConfig` instances. `TokenConfig` contains `mandatory` property, which is used to filter `minimum_tokens` out of entire list. As a byproduct, `--minimum-tokens` are automatically added to list specified by `--tokens`.
- renamed `--token-configs` to just `--config`. It's a bit misleading, as it governs only `SETTINGS['tokens']`. Should flag name be changed, or YAML file structure changed to mirror this of SETTINGS?
- tokens of type `choice` have `casesensitive` key, which tells if values of `choices` should be matched in case-sensitive way (default) or not. I'm pretty sure everyone wants to have `casesensitive: False`, but I prefer to have this configurable anyway
- improved docs and tests

Well, it turned out to be a little larger than I expected. But I believe this covers everything that we need in Satellite, then some, and creates nice foundation for further improvements.

### CODE REVIEW

**Comment by mirzal** [comment]
Also, haven't tested on Python2. I'm pretty sure it will require some changes to maintain compatibility. The question is: do we care?

**Comment by elyezer** [comment]
@mirzal not sure if Python 2 compatibility still required. We can make this python 3 only and release a specific version for that. The old version will continue available and anyone that still on legacy python can use the old version

File: testimony/config.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    def update(self, new_values):
+        """Update token configuration with dictionary of new values."""
+        for key, value in new_values.items():
+            setattr(self, key, value)
+
+    def validate(self, what):
+        """Ensure that 'what' meets value validation criteria."""
+        if self.token_type == 'choice':
+            return what.lower() in self.choices
+        return True  # assume valid for unknown types
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: If `TokenConfig.mandatory` is True, then this token name will be included in `minimum_tokens` list that is passed in to `DocstringParser`. This way `testcase.has_minimum_tokens` property will be set to False if testcase doesn't contain all required tokens. We check it in `__init__.py:356` and report it under "Test cases missing minimal docstrings".

I can agree that DocstringParser shouldn't be responsible for validating test case metadata tokens, but I tried to not rewrite the whole thing ;) .

File: docs/index.rst [hunk 2]
```diff
@@ -292,6 +301,58 @@ Misc Options
...
+single map (equivalent of Python dict), where keys are names of tokens
+and values are maps consisting of ``mandatory``, ``type`` and other,
+type-dependant keys. Sample config files are provided in ``tests``
+directory, as well as printed below:
+
+.. code-block:: console
+
+    ---
+    Assert:
+        mandatory: True  # 'Assert' is required in each test
```
[the remark this answers is not in the document]
Comment (mirzal) [reply docs/index.rst]: I think I've seen it both ways. I don't have strong opinion either way, so changed it in f926c16493aeb45eda1b37ba4cae7e36e3be4e23

File: docs/index.rst [hunk 3]
```diff
@@ -292,6 +301,58 @@ Misc Options
...
+    Feature:
+        mandatory: True
+    Test:
+        mandatory: True
+    # You can specify that token is not required explicitly, or leave it
+    # out - testimony will assume default value of 'False'
+    # Both tokens below are allowe, but not required
+    BZ:
+        mandatory: False
+    Setup: ""
```
[the remark this answers is not in the document]
Comment (mirzal) [reply docs/index.rst]: Didn't know YAML can do this :) . Changed.

However, I think we still need to validate input value in `config.py:TokenConfig:__init__`, just in case someone prepares not-exactly-correct config file. Right now it invokes `get()` method on value passed in and falls back to empty dict if there is exception.

File: testimony/config.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# -*- coding: UTF-8 -*-
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: Changed. Also changed `docs/conf.py`, `tests/sample_pkg/test_sample2.py` and `tests/test_sample.py`, which had the same `# -*- coding: UTF-8 -*-`.

File: testimony/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (elyezer) [reply testimony/config.py]: What I am saying here is to check if the value is not empty, not to check if the minimum set of tokens is available. Does it make sense?

File: testimony/cli.py [hunk 5]
```diff
@@ -12,15 +12,18 @@
 @click.option('--tokens', help='Comma separated list of expected tokens')
 @click.option(
     '--minimum-tokens', help='Comma separated list of minimum expected tokens')
+@click.option(
+    '--config', 'config_file', help='Configuration file (YAML)')
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/cli.py]: Thanks for letting me know. Done in 491d69a45270416af74ff57166152784ae0a2681 .

File: testimony/config.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import sys
+import yaml
+
+TOKEN_TYPES = ['choice']
+
+
+def parse_config(filename):
+    """Parse the config."""
+    try:
+        with open(filename, 'r') as fp:
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: Also done in 491d69a45270416af74ff57166152784ae0a2681

File: testimony/config.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            data = yaml.load(fp, Loader=yaml.SafeLoader)
+            return {k.lower(): TokenConfig(k, v) for k, v in data.items()}
+    except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: Also done in 491d69a45270416af74ff57166152784ae0a2681

File: testimony/config.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        Takes name of the token and actual config structure as a param
+        """
+        try:
+            config.get('key')
+        except AttributeError:
+            config = {}
+
+        self.name = name.lower()
+        self.mandatory = config.get('mandatory', False)
+        self.token_type = ''
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: Good point. I like `None` better, as it is less limiting for us. If we set default to `'string'`, and then decide that strings should always have some additional option, we would break backwards compatibility. There are no problems like that with `None`.

And once we are sure we don't want any required option for `'string'`, ever, we can just change it.

File: testimony/config.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        if 'type' in config and config['type'] in TOKEN_TYPES:
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: Your proposal is so much better. I like it, changed.

File: testimony/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/config.py]: I'm not sure if I understand. You mean to add support for config files like that:
```yaml
status:
    required:
```
So, to handle scenarios when `required` key is present, but doesn't have value assigned?

We already do that :) . If key is present, but doesn't have value, then it will be `None`. When we generate `minimum_tokens` list, we cast value of `required` to boolean. So key without value will be False, the same as empty list or dict.

File: tests/config-full.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
+---
+Assert:
```
[the remark this answers is not in the document]
Comment (mirzal) [reply tests/config-full.yaml]: Changed.

But I want to point out that testimony already calls `.lower()` on tokens before doing anything useful for them, so it doesn't really matter how we write them. And with mixed case you can improve readability sometimes, like with `CaseImportance` vs `caseimportance`.

File: tests/config-basic.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+---
+Assert:
```
[the remark this answers is not in the document]
Comment (mirzal) [reply tests/config-basic.yaml]: Changed.

File: testimony/cli.py [hunk 12]
```diff
@@ -12,15 +12,18 @@
 @click.option('--tokens', help='Comma separated list of expected tokens')
```
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/cli.py]: I would keep them around.

First, it's in line with what many other tools do, where you can use either config file or options passed in through command line.

Second, it's good for ad-hoc tests and experimentation. If I would like to get overview of tests that are missing `BZ` token, it's faster to run `testimony --minimum-tokens bz validate file` than to create new config file and run `testimony --config path/to/config.yaml validate file`.

File: testimony/cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mirzal) [reply testimony/cli.py]: `-c` support added.

Regarding default value: looks like [`click.File` doesn't support default argument](https://click.palletsprojects.com/en/7.x/api/#click.File)? So if we want to fall back to default, we need to switch back to handling file opening ourselves.

Another thing is that global per-user config has limited utility for linters, even more so for testimony. As you can work on multiple projects at once, and each of them will provide their own config file, you will want to use per-project configuration anyway. Sure, we could try to find config file in current directory, or any of parent directories, but that only adds complexity.

To sum up, I would create issue for that and leave it for another PR.

File: docs/index.rst [hunk 13]
```diff
@@ -292,6 +301,58 @@ Misc Options
...
+list of supported tokens (there is no need to specify single token
+in both ``--minimum-tokens`` and ``--tokens``).
+
+``--config`` is path to YAML configuration file. YAML file should contain
+single map (equivalent of Python dict), where keys are names of tokens
+and values are maps consisting of ``mandatory``, ``type`` and other,
+type-dependant keys. Sample config files are provided in ``tests``
+directory, as well as printed below:
+
+.. code-block:: console
```
[the remark this answers is not in the document]
Comment (mirzal) [reply docs/index.rst]: You mean something like `example-config.yaml` in root space of testimony, or maybe `docs`?

We don't actually build docs for testimony, but I would keep it there for now. It's probably the most discoverable place at the moment. This will become problem as we gain support for more token types, but I would just deal with it then.

**Comment by mirzal** [comment]
@elyezer I included changes that you proposed. Also replied to your comments. Please let me know if you would like to change anything else.

By the way, you are OK with merging this PR as-is, or would you like me to squash some commits? I would at least keep one commit per author.

File: docs/index.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (elyezer) [reply docs/index.rst]: Sure, just mentioned to have that in mind

**Comment by elyezer** [comment]
@mirzal can you squash please? It is fine leaving one commit per author. Thank you for all the changes.

**Comment by mirzal** [comment]
@elyezer I squashed into two commits, one per author. Commit messages might leave something to desire, but these are really one big change and I don't see much benefit in trying to make these commits atomic.

Are we OK to merge this? Can you release new version with these changes? It would be Python 3-only - as expected, testimony throws exception on Python 2.

**Comment by elyezer** [comment]
@mirzal I am going to merge this and work on making Testimony Python 3 only and fix travis. With that I will be releasing version 2.0.0 since it is going to break compatibility. I will try to get this done ASAP.

**Comment by jhutar** [comment]
Cool!
